src/lib/provable/gadgets/sha2.ts

```diff
@@ -0,0 +1,621 @@
+ * @param length Whether this is a SHA2-224 or SHA2-256 or SHA2-384 or SHA2-512
+ * @param M - The 512-bit message block (16-element array of UInt32)
+ *            or the 1024-bit message block (16-element array of UInt64).
+ * @returns The message schedule (64-element array of UInt32 or 80-element array of UInt64).
+ */
+function messageSchedule<T extends UInt32 | UInt64>(
+  length: Length,
+  M: T[]
+): T[] {
+  // §6.2.2.1 and §6.4.2.1
+
+  // Declare W as an empty array of type T[] (generic array)
+  const W: T[] = [];
+
+  // for each message block of 16 x 32bit | 64bit do:
+
+  let numWords = SHA2Constants.NUM_WORDS[length];
+
+  // prepare message block
+  for (let t = 0; t < 16; t++) W[t] = M[t];
+  for (let t = 16; t < numWords; t++) {
+    // the field element is unreduced and not proven to be 32bit | 64bit,
+    // we will do this later to save constraints
+    let unreduced = DeltaOne(length, W[t - 2])
+      .value.add(W[t - 7].value)
+      .add(DeltaZero(length, W[t - 15]).value.add(W[t - 16].value));
+
+    // mod 32 | 64 bit the unreduced field element
+    W[t] = reduceMod(length, unreduced);
+  }
+
+  return W;
+}
+
+/**
+ * Performs the SHA-2 compression function on the given hash values and message schedule.
+ *
+ * @param length Whether this is a SHA2-224 or SHA2-256 or SHA2-384 or SHA2-512
+ * @param H - The initial or intermediate hash values (8-element array of T).
+ * @param W - The message schedule (64-element array of T).
+ *
+ * @returns The updated intermediate hash values after compression.
+ */
+function compression<T extends UInt32 | UInt64>(
+  length: Length,
+  [...H]: T[],
+  W: T[]
+) {
+  let numWords = SHA2Constants.NUM_WORDS[length];
+
+  let k = SHA2Constants.K[length];
+
+  // §6.2.2.2 and §6.4.2.2:
+  // initialize working variables
+  let a = H[0];
+  let b = H[1];
+  let c = H[2];
+  let d = H[3];
+  let e = H[4];
+  let f = H[5];
+  let g = H[6];
+  let h = H[7];
+
+  // §6.2.2.3 and §6.4.2.3:
+  // main loop
+  for (let t = 0; t < numWords; t++) {
+    // T1 is unreduced and not proven to be 32 | 64 bit,
+    // we will do this later to save constraints
+    const unreducedT1 = h.value
+      .add(SigmaOne(length, e).value)
+      .add(Ch(length, e, f, g).value)
+      .add(k[t])
+      .add(W[t].value)
+      .seal();
+
+    // T2 is also unreduced
+    const unreducedT2 = SigmaZero(length, a).value.add(
+      Maj(length, a, b, c).value
+    );
+
+    h = g;
+    g = f;
+    f = e;
+    e = reduceMod(length, d.value.add(unreducedT1));
+    // mod 32 | 64 bit the unreduced field element
+    d = c;
+    c = b;
+    b = a;
+    a = reduceMod(length, unreducedT2.add(unreducedT1));
+    // mod 32 | 64 bit
+  }
+
+  // §6.2.2.4 and §6.4.2.4
+  // new intermediate hash value
+  intermediateHash(length, [a, b, c, d, e, f, g, h], H);
+
+  return H;
+}
+
+// Helper functions
+
+// Helper function to check if it is a short SHA2 (SHA2-224 or SHA2-256) or not
+function isShort(length: Length): boolean {
```

Comment:
Nit: I would inline this check as `isShort` does not have a clear semantics for the primitive.

Reply:
Yeah I didn't like it either... At first it was is256 but then it was true also in 224. Any suggestion?